C++ code initialization globals emitted before types are fully declared

Compiling the following sample currently fails:

```.ruby
# c.hlt
module c {

public type a = struct {};

global a x;

}
```

```.console
$ hiltic -j c.hlt
/private/var/folders/ht/vtb0rkbd4ws15vx02xdb84zw0000gn/T/hilti-rt-test-corFCB/c.cc:18:11: error: field has incomplete type '__hlt::c::a'
        a x{};
          ^
/private/var/folders/ht/vtb0rkbd4ws15vx02xdb84zw0000gn/T/hilti-rt-test-corFCB/c.cc:9:12: note: forward declaration of '__hlt::c::a'
    struct a;
           ^
In file included from /private/var/folders/ht/vtb0rkbd4ws15vx02xdb84zw0000gn/T/hilti-rt-test-corFCB/c.cc:1:
In file included from /Users/bbannier/src/spicy/hilti/runtime/include/hilti/rt/libhilti.h:9:
In file included from /Users/bbannier/src/spicy/build/include/hilti/rt/autogen/config.h:5:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/c++/v1/string:504:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/c++/v1/string_view:175:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/c++/v1/__string:57:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/c++/v1/algorithm:643:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/c++/v1/memory:3866:24: error: assigning to 'std::__1::__shared_weak_count *' from incompatible type
      'std::__1::__shared_ptr_emplace<__hlt::c::__globals_t, std::__1::allocator<__hlt::c::__globals_t> > *'
        __r.__cntrl_ = __cntrl;
                       ^~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/c++/v1/memory:4408:29: note: in instantiation of function template specialization
      'std::__1::shared_ptr<__hlt::c::__globals_t>::__create_with_control_block<__hlt::c::__globals_t, std::__1::__shared_ptr_emplace<__hlt::c::__globals_t, std::__1::allocator<__hlt::c::__globals_t> > >' requested here
    return shared_ptr<_Tp>::__create_with_control_block(__ptr, __hold2.release());
                            ^
/Users/bbannier/src/spicy/hilti/runtime/include/hilti/rt/global-state.h:124:59: note: in instantiation of function template specialization 'std::__1::make_shared<__hlt::c::__globals_t>' requested here
    context::detail::current()->hilti_globals[idx] = std::make_shared<T>();
                                                          ^
2 errors generated.
[error] hiltic: JIT compilation failed
```

This seems to be due to the order of the emitted definitions in the generated C++ code:

```.cpp
// ...

namespace __hlt::c {
    struct a;
}

// ...

namespace __hlt::c {
    struct __globals_t : hilti::rt::trait::isStruct, hilti::rt::Controllable<__globals_t> {
        a x{};
        template<typename F> void __visit(F _) const { _("x", x); }
    };

    struct a : hilti::rt::trait::isStruct, hilti::rt::Controllable<a> {
        inline auto __self() { return hilti::rt::ValueReference<a>::self(this); }

// ...

```

Here the definition of `__globals_t` cannot create a `a` since at that point the type has only been forward-declared.

This currently probably is not an issue for Spicy code as we wrap most values of user-specified type in some reference.

--- a/hilti/toolchain/src/compiler/codegen/codegen.cc
+++ b/hilti/toolchain/src/compiler/codegen/codegen.cc
@@ -67,7 +67,7 @@ struct GlobalsVisitor : hilti::visitor::PreOrder<void, GlobalsVisitor> {
             unit->add(t);
 
             auto body = cxx::Block();
-            body.addStatement("return hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index)");
+            body.addStatement("return ::hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index)");
 
             auto body_decl = cxx::declaration::Function{
                 .result = "auto",

--- a/hilti/toolchain/src/compiler/codegen/codegen.cc
+++ b/hilti/toolchain/src/compiler/codegen/codegen.cc
@@ -146,8 +146,12 @@ struct GlobalsVisitor : hilti::visitor::PreOrder<void, GlobalsVisitor> {
     void operator()(const declaration::Type& n, position_t p) {
         assert(n.typeID());
         cg->compile(n.type(), codegen::TypeUsage::Storage);
-        if ( include_implementation )
+        if ( include_implementation ) {
             cg->addTypeInfoDefinition(n.type());
+
+            if ( const auto& dt = cg->typeDeclaration(n.type()) )
+                cg->unit()->prioritizeType(dt->id);
+        }
     }
 };
 

--- a/hilti/toolchain/src/compiler/codegen/codegen.cc
+++ b/hilti/toolchain/src/compiler/codegen/codegen.cc
@@ -49,9 +49,6 @@ struct GlobalsVisitor : hilti::visitor::PreOrder<void, GlobalsVisitor> {
         for ( const auto& i : module.childs() )
             v.dispatch(i);
 
-        if ( v.globals.empty() && v.constants.empty() )
-            return;
-
         auto ns = cxx::ID(cg->options().cxx_namespace_intern, module_id);
 
         for ( const auto& c : v.constants )
@@ -83,7 +80,7 @@ struct GlobalsVisitor : hilti::visitor::PreOrder<void, GlobalsVisitor> {
             unit->add(body_decl);
         }
 
-        if ( include_implementation ) {
+        if ( ! v.globals.empty() && include_implementation ) {
             // Create the initGlobals() function.
             auto id = cxx::ID{ns, "__init_globals"};
 

--- a/hilti/toolchain/src/compiler/codegen/statements.cc
+++ b/hilti/toolchain/src/compiler/codegen/statements.cc
@@ -37,11 +37,11 @@ struct Visitor : hilti::visitor::PreOrder<void, Visitor> {
         std::string throw_;
 
         if ( n.message() )
-            throw_ = fmt("throw hilti::rt::AssertionFailure(hilti::rt::to_string_for_print(%s), \"%s\")",
+            throw_ = fmt("throw ::hilti::rt::AssertionFailure(hilti::rt::to_string_for_print(%s), \"%s\")",
                          cg->compile(*n.message()), n.meta().location());
         else {
             auto msg = std::string(to_node(n.expression()));
-            throw_ = fmt(R"(throw hilti::rt::AssertionFailure("failed expression '%s'", "%s"))",
+            throw_ = fmt(R"(throw ::hilti::rt::AssertionFailure("failed expression '%s'", "%s"))",
                          util::escapeUTF8(msg, true), n.meta().location());
         }
 

--- a/hilti/toolchain/src/compiler/codegen/types.cc
+++ b/hilti/toolchain/src/compiler/codegen/types.cc
@@ -54,7 +54,7 @@ struct VisitorDeclaration : hilti::visitor::PreOrder<cxx::declaration::Type, Vis
                 std::vector<cxx::type::struct_::Member> fields;
 
                 cxx::Block self_body;
-                self_body.addStatement(util::fmt("return hilti::rt::ValueReference<%s>::self(this)", id));
+                self_body.addStatement(util::fmt("return ::hilti::rt::ValueReference<%s>::self(this)", id));
 
                 auto self = cxx::declaration::Function{.result = "auto",
                                                        .id = "__self",
@@ -334,7 +334,7 @@ struct VisitorDeclaration : hilti::visitor::PreOrder<cxx::declaration::Type, Vis
         if ( sid.namespace_() )
             scope = scope.namespace_();
 
-        std::string base_ns = "hilti::rt";
+        std::string base_ns = "::hilti::rt";
         std::string base_cls = "UserException";
 
         if ( auto b = n.baseType() ) {
@@ -422,7 +422,7 @@ struct VisitorStorage : hilti::visitor::PreOrder<CxxTypes, VisitorStorage> {
 
         // Add tailored operator<<.
         auto render_body = cxx::Block();
-        render_body.addStatement("o << hilti::rt::to_string(x); return o");
+        render_body.addStatement("o << ::hilti::rt::to_string(x); return o");
 
         auto render_decl =
             cxx::declaration::Function{.result = "std::ostream&",
@@ -586,7 +586,7 @@ struct VisitorStorage : hilti::visitor::PreOrder<CxxTypes, VisitorStorage> {
             sid, [&]() { return CxxTypes{.base_type = std::string(sid)}; },
             [&](auto& cxx_types) {
                 auto render_body = cxx::Block();
-                render_body.addStatement("o << hilti::rt::to_string(x); return o");
+                render_body.addStatement("o << ::hilti::rt::to_string(x); return o");
 
                 auto render_decl =
                     cxx::declaration::Function{.result = "std::ostream&",
@@ -698,7 +698,7 @@ struct VisitorStorage : hilti::visitor::PreOrder<CxxTypes, VisitorStorage> {
             sid, [&]() { return CxxTypes{.base_type = std::string(sid)}; },
             [&](auto& cxx_types) {
                 auto render_body = cxx::Block();
-                render_body.addStatement("o << hilti::rt::to_string(x); return o");
+                render_body.addStatement("o << ::hilti::rt::to_string(x); return o");
 
                 auto render_decl =
                     cxx::declaration::Function{.result = "std::ostream&",

--- a/hilti/toolchain/src/compiler/cxx/elements.cc
+++ b/hilti/toolchain/src/compiler/cxx/elements.cc
@@ -395,8 +395,8 @@ std::string cxx::type::Struct::str() const {
     if ( args.size() )
         has_params = ", hilti::rt::trait::hasParameters";
 
-    return fmt("struct %s : hilti::rt::trait::isStruct%s, hilti::rt::Controllable<%s> {\n%s\n}", type_name, has_params,
-               type_name, struct_fields_as_str);
+    return fmt("struct %s : ::hilti::rt::trait::isStruct%s, ::hilti::rt::Controllable<%s> {\n%s\n}", type_name,
+               has_params, type_name, struct_fields_as_str);
 }
 
 std::string cxx::type::Struct::inlineCode() const {

--- a/hilti/toolchain/src/compiler/cxx/unit.cc
+++ b/hilti/toolchain/src/compiler/cxx/unit.cc
@@ -228,7 +228,7 @@ void Unit::_generateCode(Formatter& f, bool prototypes_only) {
         // Write out those types first that we have in _types_in_order.
         for ( const auto& id : _types_in_order ) {
             auto i = _types.find(id);
-            if ( i != _types.end() )
+            if ( i == _types.end() )
                 continue;
 
             auto& t = i->second;

--- a/tests/Baseline/hilti.hiltic.print.globals/output
+++ b/tests/Baseline/hilti.hiltic.print.globals/output
@@ -7,13 +7,13 @@
 #include <hilti/rt/libhilti.h>
 
 namespace __hlt::Foo {
-    struct __globals_t : hilti::rt::trait::isStruct, hilti::rt::Controllable<__globals_t> {
+    struct __globals_t : ::hilti::rt::trait::isStruct, ::hilti::rt::Controllable<__globals_t> {
         std::string X{};
         template<typename F> void __visit(F _) const { _("X", X); }
     };
 
     inline unsigned int __globals_index;
-    static inline auto __globals() { return hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index); }
+    static inline auto __globals() { return ::hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index); }
     extern void __init_globals(::hilti::rt::Context* ctx);
     extern void __init_module();
     extern void __register_module();

--- a/tests/Baseline/hilti.hiltic.print.import/output
+++ b/tests/Baseline/hilti.hiltic.print.import/output
@@ -7,24 +7,24 @@
 #include <hilti/rt/libhilti.h>
 
 namespace __hlt::Bar {
-    struct __globals_t : hilti::rt::trait::isStruct, hilti::rt::Controllable<__globals_t> {
+    struct __globals_t : ::hilti::rt::trait::isStruct, ::hilti::rt::Controllable<__globals_t> {
         std::string bar{};
         template<typename F> void __visit(F _) const { _("bar", bar); }
     };
 
     inline unsigned int __globals_index;
-    static inline auto __globals() { return hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index); }
+    static inline auto __globals() { return ::hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index); }
 }
 
 namespace __hlt::Foo {
-    struct __globals_t : hilti::rt::trait::isStruct, hilti::rt::Controllable<__globals_t> {
+    struct __globals_t : ::hilti::rt::trait::isStruct, ::hilti::rt::Controllable<__globals_t> {
         std::string foo{};
         template<typename F> void __visit(F _) const { _("foo", foo); }
     };
 
     inline unsigned int __globals_index;
     static inline auto __globals() {
-        return hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index);
+        return ::hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index);
     }
 
     extern void __init_globals(::hilti::rt::Context* ctx);
@@ -62,27 +62,27 @@ extern void __hlt::Foo::__register_module() { ::hilti::rt::detail::registerModul
 #include <hilti/rt/libhilti.h>
 
 namespace __hlt::Bar {
-    struct __globals_t : hilti::rt::trait::isStruct, hilti::rt::Controllable<__globals_t> {
+    struct __globals_t : ::hilti::rt::trait::isStruct, ::hilti::rt::Controllable<__globals_t> {
         std::string bar{};
         template<typename F> void __visit(F _) const { _("bar", bar); }
     };
 
     inline unsigned int __globals_index;
-    static inline auto __globals() { return hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index); }
+    static inline auto __globals() { return ::hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index); }
     extern void __init_globals(::hilti::rt::Context* ctx);
     extern void __init_module();
     extern void __register_module();
 }
 
 namespace __hlt::Foo {
-    struct __globals_t : hilti::rt::trait::isStruct, hilti::rt::Controllable<__globals_t> {
+    struct __globals_t : ::hilti::rt::trait::isStruct, ::hilti::rt::Controllable<__globals_t> {
         std::string foo{};
         template<typename F> void __visit(F _) const { _("foo", foo); }
     };
 
     inline unsigned int __globals_index;
     static inline auto __globals() {
-        return hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index);
+        return ::hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index);
     }
 
 }

--- a/tests/Baseline/hilti.types.real.nops/output
+++ b/tests/Baseline/hilti.types.real.nops/output
@@ -15,14 +15,14 @@ assert 0x1.999999999999ap-4 == 0x1.999999999999ap-4;
 #include <hilti/rt/libhilti.h>
 
 namespace __hlt::Foo {
-    struct __globals_t : hilti::rt::trait::isStruct, hilti::rt::Controllable<__globals_t> {
+    struct __globals_t : ::hilti::rt::trait::isStruct, ::hilti::rt::Controllable<__globals_t> {
         double d{};
         double r{};
         template<typename F> void __visit(F _) const { _("d", d); _("r", r); }
     };
 
     inline unsigned int __globals_index;
-    static inline auto __globals() { return hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index); }
+    static inline auto __globals() { return ::hilti::rt::detail::moduleGlobals<__globals_t>(__globals_index); }
     extern void __init_globals(::hilti::rt::Context* ctx);
     extern void __init_module();
     extern void __register_module();
@@ -38,7 +38,7 @@ extern void __hlt::Foo::__init_globals(::hilti::rt::Context* ctx) {
 extern void __hlt::Foo::__init_module() {
       __location__("<...>/nops.hlt:11:1");
     if ( ! (0x1.999999999999ap-4 == 0x1.999999999999ap-4) ) {
-        throw hilti::rt::AssertionFailure("failed expression '0x1.999999999999ap-4 == 0x1.999999999999ap-4'", "<...>/nops.hlt:11:1");
+        throw ::hilti::rt::AssertionFailure("failed expression '0x1.999999999999ap-4 == 0x1.999999999999ap-4'", "<...>/nops.hlt:11:1");
     }
 }
 

--- a/tests/hilti/codegen/globals-deps.hlt
+++ b/tests/hilti/codegen/globals-deps.hlt
@@ -1,0 +1,11 @@
+# @TEST-EXEC: hiltic -j %INPUT
+#
+# @TEST-DOC: This is a regression test for https://github.com/zeek/spicy/issues/1030.
+
+module c {
+
+public type a = struct {};
+
+global a x;
+
+}

